File: 01Singleton/Students/2019/OnychAleksandra/main/SingletonExample.java
package main;

import java.io.*;

public class SingletonExample implements Serializable {
    private static SingletonExample instance;
    private String message = "Message";

    private SingletonExample() { }

    public static SingletonExample getInstance() {
        if (instance == null) {
            synchronized (SingletonExample.class) {
                if (instance == null) {
                    instance = new SingletonExample();
                }
            }
        }
        return instance;
    }

    public static void cleanInstance() {
        instance = null;
    }

    protected Object readResolve() throws ObjectStreamException {
        if (instance == null)
            instance = this;
        instance.setMessage(this.getMessage());
        return instance;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }
}

File: 01Singleton/Students/2019/PiotrKawa/src/main/kotlin/SingletonExample.kt
import java.io.Serializable

class SingletonExample private constructor() : Serializable {

    companion object {
        @JvmStatic private var singleton: SingletonExample? = null

        @Synchronized @JvmStatic fun getInstance() : SingletonExample {
            if (singleton == null) { singleton = SingletonExample() }
            return singleton!!
        }

    }

    init {
        if (singleton != null) {
            throw InstantiationError("This object already has one instance created")
        }
    }

    fun readResolve() = singleton as Any
}